astroquery/utils/commons.py

```diff
@@ -389,22 +389,22 @@ def save_fits(self, savepath, link_cache='hard'):
             If the system is unable to create a hardlink, the file will be
             copied to the target location.
         """
-        from warnings import warn
-
         self.get_fits()
 
-        try:
+        target_key = str(self._target)
+        if not aud.is_url_in_cache(target_key):
+            raise IOError("Cached file not found / does not exist.")
+
+        # There has been some internal refactoring in astropy.utils.data
+        # so we do this check. See https://github.com/astropy/astropy/pull/10437
+        if hasattr(aud, '_get_download_cache_locs'):
+            import shelve
             dldir, urlmapfn = aud._get_download_cache_locs()
-        except (IOError, OSError) as e:
-            msg = 'Remote data cache could not be accessed due to '
-            estr = '' if len(e.args) < 1 else (': ' + str(e))
-            warn(aud.CacheMissingWarning(msg + e.__class__.__name__ + estr))
-
-        with _open_shelve(urlmapfn, True) as url2hash:
-            if str(self._target) in url2hash:
-                target = url2hash[str(self._target)]
-            else:
-                raise IOError("Cached file not found / does not exist.")
+            with shelve.open(urlmapfn) as url2hash:
+                target = url2hash[target_key]
+        else:
+            dldir = aud._get_download_cache_loc()
+            target = os.path.join(dldir, aud._url_to_dirname(target_key), "contents")
```

Comment:
This still requires unnecessary reference to cache internals. Try:

```python
target = download_file(target_key, cache=True, sources=[])
```
This will raise an exception if `self._target` is not in the cache, which could be caught and re-raised in a more helpful form if desired, or just protected by an if statement. `target` is guaranteed to be an absolute path to the file.

Reply:
If you want it to also work against pre-4 astropy, you can omit the `sources` argument and just use the `is_url_in_cache` as a guard.

Reply:
Thanks! I knew there must be a better way and you would know. I'll give that a try.